diff --git a/CMakeLists.txt b/CMakeLists.txt
index eb7b04b6..076a89b9 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -369,10 +369,12 @@ endif ( WIN32 )
 
 find_package ( GCEM REQUIRED )
 
+if ( NOT osal STREQUAL "embedded" )
 find_library ( HAS_LIBM NAMES "m" )
 if ( HAS_LIBM )
   set ( MATH_LIBRARY "m" )
 endif ( HAS_LIBM )
+endif ( NOT osal STREQUAL "embedded" )
 
 set ( LIBFLUID_LIBS ${MATH_LIBRARY} )
 if (NOT ((CMAKE_SYSTEM_NAME MATCHES "SunOS") OR (osal STREQUAL "embedded")))
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 388e8484..7d3ecd2a 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -180,23 +180,23 @@ set ( libfluidsynth_SOURCES
     synth/fluid_tuning.h
     synth/fluid_voice.c
     synth/fluid_voice.h
     midi/fluid_midi.c
     midi/fluid_midi.h
-    midi/fluid_midi_router.c
-    midi/fluid_midi_router.h
-    midi/fluid_seqbind.c
-    midi/fluid_seqbind_notes.cpp
-    midi/fluid_seq.c
-    midi/fluid_seq_queue.cpp
-    drivers/fluid_adriver.c
-    drivers/fluid_adriver.h
-    drivers/fluid_mdriver.c
-    drivers/fluid_mdriver.h
-    bindings/fluid_cmd.c
-    bindings/fluid_cmd.h
-    bindings/fluid_filerenderer.c
-    bindings/fluid_ladspa.c
-    bindings/fluid_ladspa.h
+    # midi/fluid_midi_router.c
+    # midi/fluid_midi_router.h
+    # midi/fluid_seqbind.c
+    # midi/fluid_seqbind_notes.cpp
+    # midi/fluid_seq.c
+    # midi/fluid_seq_queue.cpp
+    # drivers/fluid_adriver.c
+    # drivers/fluid_adriver.h
+    # drivers/fluid_mdriver.c
+    # drivers/fluid_mdriver.h
+    # bindings/fluid_cmd.c
+    # bindings/fluid_cmd.h
+    # bindings/fluid_filerenderer.c
+    # bindings/fluid_ladspa.c
+    # bindings/fluid_ladspa.h
 )
 
 set ( fluid_osal_SOURCES utils/fluid_sys_${osal}.h )
diff --git a/src/sfloader/fluid_sfont.c b/src/sfloader/fluid_sfont.c
index 7f61d63f..0ff25f4d 100644
--- a/src/sfloader/fluid_sfont.c
+++ b/src/sfloader/fluid_sfont.c
@@ -22,6 +22,7 @@
 #include "fluid_mod.h"
 
 
+#if !OSAL_embedded
 void *default_fopen(const char *path)
 {
     const char* msg;
@@ -54,6 +55,13 @@ int safe_fseek(void *handle, fluid_long_long_t ofs, int whence)
 {
     return fluid_file_seek(handle, ofs, whence);
 }
+#else
+void *default_fopen(const char *path);
+int default_fclose(void *handle);
+fluid_long_long_t default_ftell(void *handle);
+int safe_fread(void *buf, fluid_long_long_t count, void *fd);
+int safe_fseek(void *fd, fluid_long_long_t ofs, int whence);
+#endif
 
 /**
  * Creates a new SoundFont loader.
diff --git a/src/utils/fluid_settings.c b/src/utils/fluid_settings.c
index fd8cb144..1f71bf19 100644
--- a/src/utils/fluid_settings.c
+++ b/src/utils/fluid_settings.c
@@ -329,11 +329,13 @@ fluid_settings_init(fluid_settings_t *settings)
     fluid_return_if_fail(settings != NULL);
 
     fluid_synth_settings(settings);
+#if !OSAL_embedded
     fluid_shell_settings(settings);
     fluid_player_settings(settings);
     fluid_file_renderer_settings(settings);
     fluid_audio_driver_settings(settings);
     fluid_midi_driver_settings(settings);
+#endif
 }
 
 static int
diff --git a/src/utils/fluid_sys.c b/src/utils/fluid_sys.c
index 24a524a4..9d675537 100644
--- a/src/utils/fluid_sys.c
+++ b/src/utils/fluid_sys.c
@@ -224,6 +224,7 @@ fluid_log(int level, const char *fmt, ...)
     return FLUID_FAILED;
 }
 
+#if !OSAL_embedded
 void* fluid_alloc(size_t len)
 {
     void* ptr = malloc(len);
@@ -321,6 +322,7 @@ void fluid_free(void* ptr)
 {
     free(ptr);
 }
+#endif
 
 /**
  * An improved strtok, still trashes the input string, but is portable and
@@ -1586,6 +1588,7 @@ void delete_fluid_server_socket(fluid_server_socket_t *server_socket)
 
 #endif // NETWORK_SUPPORT
 
+#if !OSAL_embedded
 FILE* fluid_file_open(const char* path, const char** errMsg)
 {
     static const char ErrExist[] = "File does not exist.";
@@ -1618,6 +1621,7 @@ FILE* fluid_file_open(const char* path, const char** errMsg)
     
     return handle;
 }
+#endif
 
 fluid_long_long_t fluid_file_tell(FILE* f)
 {
diff --git a/src/utils/fluid_sys_embedded.h b/src/utils/fluid_sys_embedded.h
index be7ee9cf..638eb53a 100644
--- a/src/utils/fluid_sys_embedded.h
+++ b/src/utils/fluid_sys_embedded.h
@@ -74,9 +74,7 @@ STUB_FUNCTION(fluid_setenv, int, -1, (const char *name, const char *value, int o
 
 
 /* Time functions */
-
-STUB_FUNCTION_VOID(fluid_msleep, (unsigned int msecs))
-STUB_FUNCTION_SILENT(fluid_utime, double, 0, (void))
+/* fluid_msleep and fluid_utime provided externally by mt32-pi */
 
 
 /* Muteces */
@@ -230,7 +228,7 @@
 } fluid_stat_buf_t;
 
 STUB_FUNCTION(fluid_file_test, bool, true, (const char *path, int flags))
-STUB_FUNCTION(fluid_stat, int, -1, (const char *path, fluid_stat_buf_t *buffer))
+STUB_FUNCTION_SILENT(fluid_stat, int, -1, (const char *path, fluid_stat_buf_t *buffer))
 
 
 /* Debug functions */
diff --git a/src/utils/fluidsynth_priv.h b/src/utils/fluidsynth_priv.h
index fbbbef6b..d2a08856 100644
--- a/src/utils/fluidsynth_priv.h
+++ b/src/utils/fluidsynth_priv.h
@@ -195,13 +195,14 @@ typedef void (*fluid_rvoice_function_t)(void *obj, const fluid_rvoice_param_t pa
 
 /* Memory allocation */
 #define FLUID_MALLOC(_n)             fluid_alloc(_n)
-#define FLUID_REALLOC(_p,_n)         realloc(_p,_n)
+#define FLUID_REALLOC(_p,_n)         fluid_realloc(_p,_n)
 #define FLUID_FREE(_p)               fluid_free(_p)
 #define FLUID_NEW(_t)                (_t*)FLUID_MALLOC(sizeof(_t))
 #define FLUID_ARRAY_ALIGNED(_t,_n,_a) (_t*)FLUID_MALLOC((_n)*sizeof(_t) + ((unsigned int)_a - 1u))
 #define FLUID_ARRAY(_t,_n)           FLUID_ARRAY_ALIGNED(_t,_n,1u)
 
 void* fluid_alloc(size_t len);
+void* fluid_realloc(void* ptr, size_t len);
 
 /* File access */
 #define FLUID_FOPEN(_f,_m)           fluid_fopen(_f,_m)
